### PR TITLE
Fix deobfuscator not regenerating super maps

### LINKED ISSUE #4852: Possible obfuscation problem with Block#useNeighborBrightness (field_149783_u)
I encountered this when adding the newly-released EnderIO to the ``run/mods`` folder of an IntelliJ dev instance.  https://github.com/SleepyTrousers/EnderIO/issues/4527 has the full details, but in a nutshell:

* Tested against Forge 14.23.2.2611 and 14.23.2.2654
* EnderIO loads fine outside a dev instance
* Inside a dev instance, the instance crashes in pre-init with a ``NoSuchFieldError`` for ``field_149783_u`` - ``Block#useNeighborBrightness``.  This crash occurs with a minimal mod framework and no other mods in ``run/mods``.

See https://github.com/SleepyTrousers/EnderIO/files/1874963/crash.txt for the stacktrace.

### PR DESCRIPTION
**Pull request by bs2609**
Fixes #4852.

Changes the check in `FMLDeobfuscatingRemapper.mergeSuperMaps` to use the field map, rather than the method map. This is because while entries are inserted into both maps together, field map entries can also be removed - see here:
https://github.com/MinecraftForge/MinecraftForge/blob/bad83a303e6770d484aad0b0d891549dc273cf05/src/main/java/net/minecraftforge/fml/common/asm/transformers/deobf/FMLDeobfuscatingRemapper.java#L271-L285

### CODE REVIEW

**Comment by mezz** [comment]
Can you elaborate on this a bit more?
I know generally what this class does but it's undocumented and a bit obscure, so it will be quicker to review with some guidance.

**Comment by bs2609** [comment]
For the case of the block mentioned in the linked issue, `methodNameMaps.containsKey(parentThing)` is true, but `fieldNameMaps.containsKey(parentThing)` is false.

This means that `findAndMergeSuperMaps(parentThing)` is not called. Then, when the following code runs
```java
if (methodNameMaps.containsKey(parentThing))
{
	methodMap.putAll(methodNameMaps.get(parentThing));
}
if (fieldNameMaps.containsKey(parentThing))
{
	fieldMap.putAll(fieldNameMaps.get(parentThing));
}
```
as `fieldNameMaps.containsKey(parentThing)` is false, the field map will not inherit entries from the parent class.

This causes `Block` fields referenced in the class to not be deofuscated, which causes a `NoSuchFieldError` to occur.

**Comment by mezz** [comment]
Thanks, I think I am missing the bigger picture so I will ask some more specific questions.

- What is `findAndMergeSuperMaps` doing?
- Why are the keys in `methodNameMaps` a superset of the keys in `fieldNameMaps`? I do not understand why they are related or why nothing gets removed from `methodNameMaps`.
- Why is the block in the linked issue contained in one but not the other? Is it a field that has a name that matches an obfuscated method of a parent class?

**Comment by HenryLoenwind** [comment]
Yeah, you start with "`For the case of the block mentioned in the linked issue, methodNameMaps.containsKey(parentThing) is true, but fieldNameMaps.containsKey(parentThing) is false.`"---but I also would love to know *why* that is. (For the class in question, it cannot deobf ANY field names but method names are fine.)

**Comment by BrainStone** [comment]
Any udates?

**Comment by bs2609** [comment]
Will try to address a couple of the points mentioned.

The "top level" map keys are class names. The corresponding values are maps of (original -> remapped) field/method names.

`findAndMergeSuperMaps` is called for a class if that class does not have a field/method map associated with it. This will cause those maps to be generated.

As mentioned in the first comment, classes can have a method map entry without a field entry in the case that the class is added to both maps (from `mergeSuperMaps`), but subsequently removed from the field map (from `storeMemberFieldMapping`).

This will cause subclasses to not inherit that classes' field mappings, as they will not be regenerated.

**Comment by HenryLoenwind** [comment]
It is really hard to follow your explanation, but I guess it boils down to this: If a class doesn't access any obfuscated fields but has obfuscated methods, sub-classes of it that do access inherited obfuscated fields will fail to de-obfuscate. Correct?

**Comment by Keridos** [comment]
This seems to create a few bugs of which some are fairly bad. Any updates on this?

**Comment by LexManos** [comment]
This class, and related classes need a major rewrite. Which it will get in 1.13.
But for now this hack fix should be fine.
